### PR TITLE
add warmup inside load state dict

### PR DESCRIPTION
**Pull request by Hank0626**
增加了在graph load的时候提前推理一次的操作

### CODE REVIEW

File: python/oneflow/nn/graph/graph.py [hunk 1]
```diff
@@ -1151,6 +1151,13 @@ def _load_list_from_state_dict(state_dict):
                 tensor_list.append(tensor_of_item.to(device_of_item))
             return (name_list, convert_to_tensor_tuple(tensor_list))
 
+        def _load_input_from_state_dict(state_dict):
```
Comment (strint) [inline python/oneflow/nn/graph/graph.py:1154]: 改成 `__load_input_from_runtime_state_dict` 吧

File: python/oneflow/nn/graph/graph.py [hunk 2]
```diff
@@ -1151,6 +1151,13 @@ def _load_list_from_state_dict(state_dict):
...
+            tensor_list = []
+            for _, item in state_dict.items():
+                tensor_of_item, device_of_item = item
+                tensor_list.append(tensor_of_item.to(device_of_item))
+            return tuple(tensor_list)
```
Comment (strint) [inline python/oneflow/nn/graph/graph.py:1159]: 这个实现是不通用的， input 不一定是 tuple

Comment (Hank0626) [reply python/oneflow/nn/graph/graph.py:1159]: 那要如何把这三个input输入进去呢？

Comment (strint) [reply python/oneflow/nn/graph/graph.py:1159]: 参考 output 的恢复来写一个吧

```
_eager_outputs_index = state_dict["outputs_original"]
        def get_tensor_in_tuple(map_item):
            if isinstance(map_item, str) and map_item.startswith("_OFTPI"):
                of_idx = int(map_item[6:])
                return self._outputs_tensor_tuple[of_idx]
            else:
                return map_item
        _eager_outputs, _ = self.__map_io_lite(
            get_tensor_in_tuple, *_eager_outputs_index
        )
```

File: python/oneflow/nn/graph/graph.py [hunk 3]
```diff
@@ -1229,6 +1236,11 @@ def get_tensor_in_tuple(map_item):
         self._c_nn_graph.align_states_after_logical_graph_compile()
         self._c_nn_graph.init_runtime()
         self._is_compiled = True
+<<<<<<< HEAD
+        self.__run(__load_input_from_runtime_state_dict(state_dict["inputs"]))
```
Comment (strint) [inline python/oneflow/nn/graph/graph.py:1240]: 刚评论错了，名字可以不用改，需要改实现

Comment (Hank0626) [reply python/oneflow/nn/graph/graph.py:1239]: 我已经改了，也没关系吧

File: python/oneflow/nn/graph/graph.py [hunk 4]
```diff
@@ -1157,17 +1167,32 @@ def _load_list_from_state_dict(state_dict):
...
 
-        def get_tensor_in_tuple(map_item):
+        def get_input_tensor_in_tuple(map_item):
+            if isinstance(map_item, str) and map_item.startswith("_OFTPI"):
+                of_idx = int(map_item[6:])
+                return self._inputs_tensor_tuple[of_idx]
+            else:
+                return map_item
+
+        def get_output_tensor_in_tuple(map_item):
```
Comment (strint) [inline python/oneflow/nn/graph/graph.py:1182]: get_input_tensor_in_tuple 和 get_output_tensor_in_tuple 的实现一样，应该可以复用

File: python/oneflow/nn/graph/graph.py [hunk 5]
```diff
@@ -954,6 +954,7 @@ def _compile_from_shared(self, *args, **kwargs):
         inputs_tensor_tuple = convert_to_tensor_tuple(
             self.__flatten_io("input", *args, **kwargs)
         )
+        self.inputs_original = (args, kwargs)
```
Comment (strint) [inline python/oneflow/nn/graph/graph.py:957]: 传递到 __prepare_for_share_or_runtime_save 这个函数里面做 save 吧，然后使用下里面的判断条件：在 enable save 的时候才保存 inputs_original

File: python/oneflow/nn/graph/graph.py [hunk 6]
```diff
@@ -1472,6 +1498,7 @@ def __build_graph(self, *args, **kwargs):
             inputs_tensor_tuple = convert_to_tensor_tuple(
                 self.__flatten_io("input", *args, **kwargs)
             )
+            self.inputs_original = (args, kwargs)
```
Comment (strint) [inline python/oneflow/nn/graph/graph.py:1501]: 传递到 __prepare_for_share_or_runtime_save 这个函数里面做 save 吧

File: python/oneflow/nn/graph/graph.py [hunk 7]
```diff
@@ -1229,6 +1254,7 @@ def get_tensor_in_tuple(map_item):
         self._c_nn_graph.align_states_after_logical_graph_compile()
         self._c_nn_graph.init_runtime()
         self._is_compiled = True
+        self.__run(*_eager_inputs_args, **_eager_inputs_kwargs)  # pre-run to warm up
```
Comment (strint) [inline python/oneflow/nn/graph/graph.py:1257]: 给 load_runtime_state_dict 增加一个 bool 变量， warmup_with_run=False

用 warmup_with_run 来控制是否执行 self.__run

File: python/oneflow/nn/graph/cache.py [hunk 8]
```diff
@@ -175,7 +175,9 @@ def _init_and_get_a_graph_in_cache(self, cache_key):
         return graph
 
     def load_runtime_state_dict(
-        self, state_dict: Dict[str, Dict[str, Union[Dict[str, Tensor], str]]]
+        self,
+        state_dict: Dict[str, Dict[str, Union[Dict[str, Tensor], str]]],
+        warmup_with_run: bool,
```
Comment (strint) [inline python/oneflow/nn/graph/cache.py:180]: 改成这样吧
```
        self,
        state_dict: Dict[str, Dict[str, Union[Dict[str, Tensor], str]]],
        *,
        warmup_with_run: bool = False,
```
提供默认值，且要求 warmup_with_run 必须是 kwargs 形式

File: python/oneflow/nn/graph/graph.py [hunk 9]
```diff
@@ -1157,17 +1171,28 @@ def _load_list_from_state_dict(state_dict):
...
                 return map_item
 
+        _eager_inputs_args, _eager_inputs_kwargs = self.__map_io_lite(
+            partial(get_tensor_in_tuple, "input"),
+            *_eager_inputs_args_index,
+            **_eager_inputs_kwargs_index,
+        )
         _eager_outputs, _ = self.__map_io_lite(
-            get_tensor_in_tuple, *_eager_outputs_index
+            partial(get_tensor_in_tuple, "output"), *_eager_outputs_index
```
Comment (strint) [inline python/oneflow/nn/graph/graph.py:1195]: 改成 partial(get_tensor_in_tuple, self._outputs_tensor_tuple) 应该更好

**Review by strint — APPROVED** [review]
LGTM

File: python/oneflow/nn/graph/graph.py [hunk 10]
```diff
@@ -1054,20 +1056,29 @@ def _fill_sub_destination(dest_dict, name_list, tensor_tuple):
...
-        destination["inputs"] = inputs_sub_destination
-
         # This is original outputs is needed to build output buffer.
         tuple_idx = -1
 
         def gen_index_in_tuple(eager_out):
             nonlocal tuple_idx
             tuple_idx += 1
-            return "_OFTPI" + str(tuple_idx)
+            return "_OFTPI" + str(tuple_idx)  # oneflow tuple index
```
Comment (jackalcooper) [inline python/oneflow/nn/graph/graph.py:1065]: oneflow tuple index是什么意思可以写的更清楚些

File: python/oneflow/nn/graph/graph.py [hunk 11]
```diff
@@ -1157,17 +1172,26 @@ def _load_list_from_state_dict(state_dict):
...
                 return map_item
 
+        _eager_inputs_args, _eager_inputs_kwargs = self.__map_io_lite(
+            partial(get_tensor_in_tuple, self._inputs_tensor_tuple),
+            *_eager_inputs_args_index,
+            **_eager_inputs_kwargs_index,
+        )
         _eager_outputs, _ = self.__map_io_lite(
-            get_tensor_in_tuple, *_eager_outputs_index
+            partial(get_tensor_in_tuple, self._outputs_tensor_tuple),
```
Comment (jackalcooper) [inline python/oneflow/nn/graph/graph.py:1193]: 建议不用不用partial，用个lambda包一些，日后修改容易看懂

File: python/oneflow/nn/graph/graph.py (hunk 10 above)
Comment (jackalcooper) [inline python/oneflow/nn/graph/graph.py:1065]: 这段代码比较smell，是一个计数器却用了 nonlocal 还要定义一个函数

File: python/oneflow/nn/graph/graph.py [hunk 12]
```diff
@@ -1054,20 +1056,29 @@ def _fill_sub_destination(dest_dict, name_list, tensor_tuple):
...
+        )
+
+        _eager_inputs_args, _eager_inputs_kwargs = self.__map_io(
+            "input",
+            gen_index_in_tuple,
+            *self.inputs_original[0],
+            **self.inputs_original[1],
+        )
+        destination["inputs"] = inputs_sub_destination
+        destination["inputs_original"] = (_eager_inputs_args, _eager_inputs_kwargs)
```
Comment (jackalcooper) [inline python/oneflow/nn/graph/graph.py:1079]: 这个应当声明一个类或者dataclass？这里用dict不合适吧

Comment (strint) [reply python/oneflow/nn/graph/graph.py:1079]: > 这个应当声明一个类或者dataclass？这里用dict不合适吧

这个遵循了 state_dict 的习惯，是用于保存到磁盘的

File: python/oneflow/nn/graph/graph.py (hunk 10 above)
Comment (strint) [reply python/oneflow/nn/graph/graph.py:1065]: > 这段代码比较smell，是一个计数器却用了 nonlocal 还要定义一个函数

使用函数是因为调用它的接口约束了必须使用一个函数且输入是一个 tensor。所以为了 match 使用方的要求写成了这样。

Comment (strint) [reply python/oneflow/nn/graph/graph.py:1065]: > oneflow tuple index是什么意思可以写的更清楚些

背景是一个输入或者输入，是 dict/list/tuple 任意嵌套的。它有两种形式：
- 原始的；
- flatten 后的 tuple 形式的，对原始的输入做深度遍历得到的；

保存时，对原始的和 tuple 形式的都有需要，但是为了避免一个 tensor 保存两次带来的性能和正确性问题，所以原始的结构里面的 tensor 被替换成了一个 index，index 指向它在 tuple 中的位置。

全称应该是 index_in_flattened_tuple_of_this_tensor_oneflow

File: python/oneflow/nn/graph/graph.py (hunk 11 above)
Comment (Hank0626) [reply python/oneflow/nn/graph/graph.py:1193]: 意思是写成这样？
```python
_eager_inputs_args, _eager_inputs_kwargs = self.__map_io_lite(
            lambda map_item : get_tensor_in_tuple(self._inputs_tensor_tuple, map_item)
            *_eager_inputs_args_index,
            **_eager_inputs_kwargs_index,
        )
```

File: python/oneflow/nn/graph/graph.py (hunk 12 above)
Comment (jackalcooper) [reply python/oneflow/nn/graph/graph.py:1079]: 保存到磁盘没问题，但是不应该直接把orderdict数据结构暴露

Comment (strint) [reply python/oneflow/nn/graph/graph.py:1079]: > 这个应当声明一个类或者dataclass？这里用dict不合适吧

https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.state_dict

这个和 torch 的数据结构是对齐的。用 OrderedDict 比较方便遍历和扩展数据，而且支持嵌套 OrderedDict。算是惯例，也比较方便使用。

dataclass 想动态的加 attr、嵌套、遍历好像都不太方便。

File: python/oneflow/nn/graph/graph.py (hunk 11 above)
Comment (Hank0626) [reply python/oneflow/nn/graph/graph.py]: 改好了

File: python/oneflow/nn/graph/graph.py (hunk 12 above)
Comment (Hank0626) [reply python/oneflow/nn/graph/graph.py:1079]: @jackalcooper

**Review by jackalcooper — APPROVED** [review]

File: python/oneflow/nn/graph/graph.py (hunk 10 above)
Comment (strint) [reply python/oneflow/nn/graph/graph.py:1065]: @Hank0626 这段帮忙转成英文加到注释里
